realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java

```diff
@@ -3344,46 +3423,104 @@ public void distinct_async_doesNotExist() {
+                }
             }
         }
         looperThread.testComplete();
     }
 
+    // Smoke test of async distinct index linked types. Underlying mechanism is the same as for sync test
+    // (distinct_linkedAllFields), so just verifying async mechanism.
     @Test
     @RunTestInLooperThread
     public void distinct_async_indexedLinkedFields() {
         final long numberOfBlocks = 3;
         final long numberOfObjects = 3;
+        final Realm realm = looperThread.getRealm();
         populateForDistinct(realm, numberOfBlocks, numberOfObjects, false);
 
+        DynamicRealm dynamicRealm = DynamicRealm.createInstance(realm.sharedRealm);
+        RealmResults<DynamicRealmObject> all = dynamicRealm.where(AnnotationIndexTypes.CLASS_NAME)
+                .findAll();
+
+        AtomicInteger count = new AtomicInteger(0);
+
         for (String field : AnnotationIndexTypes.INDEX_FIELDS) {
-            try {
-                realm.where(AnnotationIndexTypes.class).distinct(AnnotationIndexTypes.FIELD_OBJECT + "." + field).findAllAsync();
-                fail("Unsupported " + field + " linked field");
-            } catch (IllegalArgumentException ignored) {
-            }
+            String fieldName = AnnotationIndexTypes.FIELD_OBJECT + "." + field;
+            RealmResults<AnnotationIndexTypes> distinct = realm.where(AnnotationIndexTypes.class).distinct(fieldName).findAllAsync();
+            looperThread.keepStrongReference(distinct);
+            distinct.addChangeListener(new RealmChangeListener<RealmResults<AnnotationIndexTypes>>() {
+                @Override
+                public void onChange(RealmResults<AnnotationIndexTypes> annotationIndexTypes) {
+                    Set<List<? super Object>> values = RealmQueryTests.this.distinct(all, fieldName);
+                    assertEquals(field, values.size(), distinct.size());
+                    int i = count.incrementAndGet();
+                    if (i == AnnotationIndexTypes.INDEX_FIELDS.length) {
+                        looperThread.testComplete();
+                    }
+                }
+            });
         }
-        looperThread.testComplete();
     }
 
+    // Smoke test of async distinct index linked types. Underlying mechanism is the same as for sync test
+    // (distinct_linkedAllFields), so just verifying async mechanism.
     @Test
     @RunTestInLooperThread
     public void distinct_async_notIndexedLinkedFields() {
+        final long numberOfBlocks = 3;
+        final long numberOfObjects = 3;
+        final Realm realm = looperThread.getRealm();
+        populateForDistinct(realm, numberOfBlocks, numberOfObjects, false);
+
+        DynamicRealm dynamicRealm = DynamicRealm.createInstance(realm.sharedRealm);
+        RealmResults<DynamicRealmObject> all = dynamicRealm.where(AnnotationIndexTypes.CLASS_NAME)
+                .findAll();
+
+        for (String field : AnnotationIndexTypes.NOT_INDEX_FIELDS) {
```

Comment:
Does this indexing distincting still makes sense or should it be combined with the above test case?

Reply:
As per discussion I have removed explicit differentiation of testing `distinct` on both indexed and non-indexed fields. The synchronous tests covers all types for direct, linked and nested linked objects. Async and multi argument distinct are only smoke tested for some types and unsupported types.